Import stock from a csv file

mbayopanda opened the pull request:
This PR a the option of importing stock into a depot from a csv file.

![2018-09-12_12-05-19](https://user-images.githubusercontent.com/5445251/45421326-8d751900-b684-11e8-80f0-990ce1e73469.gif)

mbayopanda committed:
import stock from file

mbayopanda committed:
create procedure for importing stock

mbayopanda committed:
importing stocks from csv file

mbayopanda committed:
handle accounting stock for importing stock

mbayopanda committed:
e2e and integration tests for stock import

mbayopanda committed:
fix linux eol

mbayopanda committed:
fix integration tests file path

jniles commented on client/src/i18n/en/stock.json:
```diff
@@ -57,6 +57,15 @@
     "PRODUCT_NOT_AVAILABLE" : "Products not available",
+    "IMPORT" : {
+      "STOCK_FROM_FILE" : "Import stock from a file",
+      "ABOUT_DESCRIPTION" : "You are about to import the stocks for the following deposit",
```
Small typo: `deposit` -> `depot`.

jniles commented on client/src/modules/stock/import/stockImport.js:
```diff
@@ -0,0 +1,108 @@
+      data : { file, depot_uuid : vm.depot.uuid },
+    };
+
+    const filters = [
```
Could you move these `filters` to outside the function?  Since they never change, there is no reason to have them in the `uploadFile()`.

jniles commented on client/src/modules/stock/import/stockImport.js:
```diff
@@ -0,0 +1,108 @@
+ * This module helps to import stock from a file
+ */
+function StockImportController(
+  Depots, Inventory, Notify, Session, util, bhConstants, ReceiptModal, StockForm,
```
It looks like there is a lot of imports here that you never use.  Can you clean this up?

jniles commented on server/controllers/inventory/import/index.js:
```diff
@@ -34,7 +34,9 @@ function downloadTemplate(req, res, next) {
 function importInventories(req, res, next) {
   if (!req.files || req.files.length === 0) {
-    next(new BadRequest('Something broke', 'ERRORS.EVERYTHING_BAD'));
+    const errorDescription = 'Expected at least one file upload but did not receive any files.';
```
This kind of error handling looks like it would be best handled by a middleware.  Could we change the `upload.middleware()` function, or create our own middleware, that simply throws an error if a route does not have `req.files` set?

jniles commented on server/controllers/stock/core.js:
```diff
@@ -496,6 +496,7 @@ function processMultipleLots(inventories) {
       lot.S_LOT_LIFETIME = lot.lifetime - lotLifetime;
       lot.S_RISK = lot.S_LOT_LIFETIME - lot.S_MONTH;
       lot.S_RISK_QUANTITY = lot.S_RISK * lot.avg_consumption;
+      lot.S_RISK_QUANTITY = Math.round(lot.S_RISK_QUANTITY);
```
It's slightly confusing to set `lot.S_RISK_QUANTITY` twice.  Could you just wrap the first line with`Math.round()`?

jniles commented on server/controllers/stock/import.js:
```diff
@@ -0,0 +1,117 @@
+
+  const filePath = req.files[0].path;
+  const depotUuid = db.bid(req.body.depot_uuid);
+  const documentUuid = db.bid(uuid.v4());
```
You use use [util.uuid()](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/lib/util.js#L248) to generate your UUIDs.

jniles commented on server/controllers/stock/import.js:
```diff
@@ -0,0 +1,117 @@
+      const transaction = db.transaction();
+
+      data.forEach(item => {
+        query = 'CALL ImportStock(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);';
```
This query never changes - you could move it out of the `data.forEach()` loop.

jniles commented on server/controllers/stock/import.js:
```diff
@@ -0,0 +1,117 @@
+  return data.every(item => {
+    return item.inventory_group_name
+      && item.inventory_text && item.inventory_type && item.inventory_unit
+      && item.inventory_unit_price && item.stock_lot_label
```
If `inventory_unit_price` is 0 ... is that allowed?

mbayopanda replied:
Zero must be allowed, so i think we have to handle that for allowing zero or to just be sure that `item.inventory_unit_price` is not undefined

jniles commented on server/models/bhima.sql:
```diff
@@ -98,7 +98,8 @@ INSERT INTO unit VALUES
   (206, '[OHADA] Bilan','TREE.OHADA_BALANCE_SHEET','',144,'/modules/reports/ohada_balance_sheet_report','/reports/ohada_balance_sheet_report'),
-  (207, 'Account Reference Report','TREE.ACCOUNT_REFERENCE_REPORT','',144,'/modules/reports/account_reference','/reports/account_reference');
+  (207, 'Account Reference Report','TREE.ACCOUNT_REFERENCE_REPORT','',144,'/modules/reports/account_reference','/reports/account_reference'),
+  (208, 'Import Stock From File','TREE.IMPORT_STOCK_FROM_FILE','',160,'/modules/stock/import','/stock/import');
```
Could you add a script to the [`migrate.sql`](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/models/migrations/v0.8.0-v0.9.0/migrate.sql) file to make these changes for us?

jniles commented on server/controllers/stock/import.js:
```diff
@@ -0,0 +1,117 @@
+          item.stock_lot_quantity,
+          moment(item.stock_lot_expiration).format('YYYY-MM-DD'),
+        ];
+        transaction.addQuery(query, queryParams);
```
I just thought of this - but most of our other modules that write multiple lines in the database use `Staging` tables ... such as [invoicing](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/models/procedures/invoicing.sql#L15), cash, and vouchers.  In theory, this gives us both a performance improvement (since we can import all the data at one time) and better error handling (since we can say that inventory XYZ doesn't exist before we even start to import the data).

It would be worth writing a second version of this function that uses temporary tables to compare with this version.  That doesn't need to be done for this PR, but we should do some performance analysis just to see.

jniles reviewed (commented):
@mbayopanda just made a few notes here.  I haven't completed a full review, but will do so later this evening after my meetings.

mbayopanda committed:
Fix and enhancements

- Fix depot english translation
- Add a middleware for handling files are presents or not
- Enhance the check of unit price by considering the zero option
- Add units in the migrate.sql file

mbayopanda committed:
remove not necessary dependencies in stockImport.js

mbayopanda committed:
fix delay default value

mbayopanda committed:
fix e2e tests dates

jeremielodi commented on server/controllers/stock/reports/stock_lots.report.handlebars:
```diff
@@ -46,8 +46,14 @@
             <th>{{translate 'STOCK.INVENTORY_UNIT'}}</th>
             <th>{{translate 'STOCK.ENTRY_DATE'}}</th>
-            <th>{{translate 'STOCK.EXPIRATION_DATE'}}</th>
+            <th>{{translate 'STOCK.EXPIRATION_DATE'}}</th>`
```
Is this cote important, if not, go head and remove it

jeremielodi reviewed (commented):
Everything looks correct!

mbayopanda committed:
remove not necessary quote
